File: typescript/src/main/vsdk/toolkit/raycasting/bidirectionalRaytracing/BidirectionalPathRaytracerConfig.ts
export class BidirectionalPathRaytracerConfig {
  public static readonly MAX_REGEXP_SIZE = 100;

  public maximumEyePathDepth: number;
  public maximumLightPathDepth: number;
  public maximumPathDepth: number;
  public minimumPathDepth: number;

  public samplesPerPixel: number;
  public totalSamples: number;
  public sampleImportantLights: number;
  public progressiveTracing: number;
  public eliminateSpikes: number;

  public useSpars: number;
  public doLe: number;
  public doLD: number;
  public doLI: number;

  public leRegExp: string;
  public ldRegExp: string;
  public liRegExp: string;

  public doWeighted: number;
  public wleRegExp: string;
  public wldRegExp: string;

  public doDensityEstimation: number;

  public constructor() {
    this.maximumEyePathDepth = 0;
    this.maximumLightPathDepth = 0;
    this.maximumPathDepth = 0;
    this.minimumPathDepth = 0;

    this.samplesPerPixel = 0;
    this.totalSamples = 0;
    this.sampleImportantLights = 0;
    this.progressiveTracing = 0;
    this.eliminateSpikes = 0;

    this.useSpars = 0;
    this.doLe = 0;
    this.doLD = 0;
    this.doLI = 0;

    this.leRegExp = "";
    this.ldRegExp = "";
    this.liRegExp = "";

    this.doWeighted = 0;
    this.wleRegExp = "";
    this.wldRegExp = "";

    this.doDensityEstimation = 0;
  }

  public copyFrom(other: BidirectionalPathRaytracerConfig | null): void {
    if (other === null) {
      return;
    }
    this.maximumEyePathDepth = other.maximumEyePathDepth;
    this.maximumLightPathDepth = other.maximumLightPathDepth;
    this.maximumPathDepth = other.maximumPathDepth;
    this.minimumPathDepth = other.minimumPathDepth;
    this.samplesPerPixel = other.samplesPerPixel;
    this.totalSamples = other.totalSamples;
    this.sampleImportantLights = other.sampleImportantLights;
    this.progressiveTracing = other.progressiveTracing;
    this.eliminateSpikes = other.eliminateSpikes;
    this.useSpars = other.useSpars;
    this.doLe = other.doLe;
    this.doLD = other.doLD;
    this.doLI = other.doLI;
    this.leRegExp = other.leRegExp;
    this.ldRegExp = other.ldRegExp;
    this.liRegExp = other.liRegExp;
    this.doWeighted = other.doWeighted;
    this.wleRegExp = other.wleRegExp;
    this.wldRegExp = other.wldRegExp;
    this.doDensityEstimation = other.doDensityEstimation;
  }
}

File: java/base/src/main/java/vsdk/toolkit/raycasting/bidirectionalRaytracing/BidirectionalPathRaytracerConfig.java
/**
Encapsulates option handling for bidirectional path tracing
*/

package vsdk.toolkit.raycasting.bidirectionalRaytracing;

/**
Contains basic config options used throughout the complete BPT code. typically two
instances are used. One persistent, containing current GUI options and one non
persistent copy when rendering an image
*/
public class BidirectionalPathRaytracerConfig {
    public static final int MAX_REGEXP_SIZE = 100;

    // Path depth config
    public int maximumEyePathDepth;
    public int maximumLightPathDepth;
    public int maximumPathDepth;
    public int minimumPathDepth;

    // Sampling details
    public int samplesPerPixel;
    public long totalSamples;
    public int sampleImportantLights;
    public int progressiveTracing;
    public int eliminateSpikes;

    // SPaR details (usage of reg exps)
    public int useSpars;
    public int doLe;
    public int doLD;
    public int doLI;

    public String leRegExp = "";
    public String ldRegExp = "";
    public String liRegExp = "";

    public int doWeighted; // Use weighted multi-pass method for a part of le and ld transport
    public String wleRegExp = "";
    public String wldRegExp = "";

    // These values are not yet in the presets or UI!
    public int doDensityEstimation;

    public void copyFrom(BidirectionalPathRaytracerConfig other) {
        if ( other == null ) {
            return;
        }
        maximumEyePathDepth = other.maximumEyePathDepth;
        maximumLightPathDepth = other.maximumLightPathDepth;
        maximumPathDepth = other.maximumPathDepth;
        minimumPathDepth = other.minimumPathDepth;
        samplesPerPixel = other.samplesPerPixel;
        totalSamples = other.totalSamples;
        sampleImportantLights = other.sampleImportantLights;
        progressiveTracing = other.progressiveTracing;
        eliminateSpikes = other.eliminateSpikes;
        useSpars = other.useSpars;
        doLe = other.doLe;
        doLD = other.doLD;
        doLI = other.doLI;
        leRegExp = other.leRegExp;
        ldRegExp = other.ldRegExp;
        liRegExp = other.liRegExp;
        doWeighted = other.doWeighted;
        wleRegExp = other.wleRegExp;
        wldRegExp = other.wldRegExp;
        doDensityEstimation = other.doDensityEstimation;
    }
}
